tfhe/examples/sha256_bool/boolean_ops.rs

```diff
@@ -0,0 +1,394 @@
+
+                } else { // grey cell
+                    new_p = propagate[index].clone();
+                    new_g = sk.or(&generate[index], &sk.and(&g, &propagate[index]));
+                }
+                (index, new_p, new_g)
+            })
+            .collect();
+
+        for (index, new_p, new_g) in updates {
+            propagate[index] = new_p;
+            generate[index] = new_g;
+        }
+    }
+
+    let mut carry = trivial_bools(&[false; 32], sk);
+    for bit in 0..31 {
+        carry[bit] = generate[bit + 1].clone();
+    }
+
+    carry
+}
+
+// 2 (homomorphic) bitwise ops
+pub fn sigma0(x: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    let a = rotate_right(x, 7, sk);
+    let b = rotate_right(x, 18, sk);
+    let c = shift_right(x, 3, sk);
+    xor(&xor(&a, &b, sk), &c, sk)
+}
+
+pub fn sigma1(x: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    let a = rotate_right(x, 17, sk);
+    let b = rotate_right(x, 19, sk);
+    let c = shift_right(x, 10, sk);
+    xor(&xor(&a, &b, sk), &c, sk)
+}
+
+pub fn sigma_upper_case_0(x: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    let a = rotate_right(x, 2, sk);
+    let b = rotate_right(x, 13, sk);
+    let c = rotate_right(x, 22, sk);
+    xor(&xor(&a, &b, sk), &c, sk)
+}
+
+pub fn sigma_upper_case_1(x: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    let a = rotate_right(x, 6, sk);
+    let b = rotate_right(x, 11, sk);
+    let c = rotate_right(x, 25, sk);
+    xor(&xor(&a, &b, sk), &c, sk)
+}
+
+// 0 bitwise ops
+fn rotate_right(x: &[Ciphertext; 32], n: usize, sk: &ServerKey) -> [Ciphertext; 32] {
+    let mut result = trivial_bools(&[false; 32], sk);
+    for i in 0..32 {
+        result[(i + n) % 32] = x[i].clone();
+    }
+    result
+}
+
+fn shift_right(x: &[Ciphertext; 32], n: usize, sk: &ServerKey) -> [Ciphertext; 32] {
+    let mut result = trivial_bools(&[false; 32], sk);
+    for i in 0..(32 - n) {
+        result[i + n] = x[i].clone();
+    }
+    result
+}
+
+// 1 bitwise op
+pub fn ch(x: &[Ciphertext; 32], y: &[Ciphertext; 32], z: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    mux(x, y, z, sk)
+}
+
+// 4 bitwise ops
+pub fn maj(x: &[Ciphertext; 32], y: &[Ciphertext; 32], z: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+
+    let (lhs, rhs) = rayon::join(
+        || and(x, &xor(y, z, sk), sk),
+        || and(y, z, sk),
+    );
+    xor(&lhs, &rhs, sk)
+}
+
+// Parallelized homomorphic bitwise ops
+// Building block for most of the previous functions
+fn xor(a: &[Ciphertext; 32], b: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    let result: Vec<Ciphertext> = (0..32)
+        .into_par_iter()
+        .map(|i| sk.xor(&a[i], &b[i]))
+        .collect();
+
+    let mut array = trivial_bools(&[false; 32], sk);
+    for (i, elem) in result.into_iter().enumerate() {
+        array[i] = elem;
+    }
```

Comment:
if the above is not relevant, something like the following avoids indexing which has range checks for each indexing op :)

```rust
array.as_mut().copy.from_slice(&result);
```

to be applied where relevant in your code